### ISSUE TITLE
A class can implement an interface with only a getter for a non-read-only field

### ISSUE THREAD

**Issue by bendmorris** [microsoft/TypeScript#49345]
# Bug Report

### 🔎 Search Terms

interface, class, getter, setter, readonly

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about read-only fields, getters and setters

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.3.3#code/JYOwLgpgTgZghgYwgAgBrIN4FgBQz-IzAQA2AJgFzIgCuAtgEbQDcuAvrrgiXAM6-IAmsmB0ADiQh0I4Aemx4CAcwhhCxcgAoAlJmRRVNKCGQBGZG1Y4OOLgHsQvNQE8qwgLx6ipSmcu5nADpvcmRPAGYre0c1AA8qdE9nK1jgjTIw5AAmKyA)

### 💻 Code

```ts
interface X {
    field: number;
}

class Y implements X {
    get field() { return 1 };
    // set field(x: number) {}
}

// this generates an error: "cannot assign to field because it is a read-only property"
const y: Y = { field: 1};
y.field = 3;

// but this is fine
const x: X = y;
x.field = 2;
```

### 🙁 Actual behavior

A class is considered to implement an interface when it includes only a getter for that interface's field; no error is generated.

### 🙂 Expected behavior

Instances of this class can't be substituted for instances of this interface, so the class definition should generate an error.

Since the interface implies that one can both get and set `field`, the class should be considered incomplete without both a getter and setter for `field`. Given a member of the interface, setting `field` typechecks, but given a member of the class, it throws an error.

The example is a minimal repro from a real bug we found that followed this pattern, and we were surprised that it was allowed.

**Comment by fatcerberus** [comment]
Not saying this isn't a bug, but note that *in general* TypeScript's readonly guarantees aren't very strong and the type system largely acts as if objects were passed by value, so this issue is just the tip of the iceberg:

```ts
interface X { field: 1 | 2; }
interface Y { readonly field: 1 | 2; }
interface Z { field: number; }

// warning: badness
const y: Y = { field: 1 };
const x: X = y;  // obj with readonly field assigned to one with writable field
x.field = 2;
console.log(y.field);  // readonly; original value was 1; prints 2

// and if that weren't bad enough...
const z: Z = y;  // obj with narrow readonly field assigned to one with wide writable field
z.field = 42;
console.log(y.field);  // type is 1 | 2; readonly; original value was 1; prints 42(!)
```

**Comment by RyanCavanaugh** [comment]
A useful mental model is that properties are "read-only" from the perspective of how the type system treats them, but allows writes anyway. There are any number of soundness holes as a result (subtyping being the obvious one, and indeed you can think of `field` as having a setter of type `never`).

**Comment by bendmorris** [comment]
Just so I'm clear - is this a known limitation that potentially will be fixed in the future, or a known limitation that we don't want to do anything about?

**Comment by fatcerberus** [comment]
> A useful mental model is that properties are "read-only" from the perspective of how the type system treats them, but allows writes anyway unless they're marked readonly.

I prefer to think of it more as the type system assuming pass-by-value semantics (i.e. assignment implies a copy is made), as this is the only way I can justify `{ readonly x: "foo" | "bar" } -> { x: string }` being a legal assignment.  And to be fair this *does* accurately model most POJO scenarios, but kind of falls apart when used to model subtyping.
